nwbwidgets/placefield.py

```diff
@@ -0,0 +1,501 @@
+import numpy as np
+from scipy.ndimage import label
+from scipy.ndimage.filters import gaussian_filter, maximum_filter
+
+import matplotlib.pyplot as plt
+import nelpy.plotting as nlp
+
+import pynwb
+from ipywidgets import widgets, BoundedFloatText, Dropdown
+
+
+from .utils.widgets import interactive_output
+from .utils.units import get_spike_times
+from .utils.timeseries import get_timeseries_in_units, get_timeseries_tt
+from .base import vis2widget
+
+
+## To-do
+# [X] Create PlaceFieldWidget class
+    # [X] Refactor place field calculation code to deal with nwb data type
+        # [X] Incorporate place field fxns into class
+        # [X] Change all internal attributes references
+        # [X]Change all internal method references
+
+    # [X] Get pos
+    # [X] Get time
+    # [X] Get spikes
+    # [] Get trials / epochs
+
+# [X] Submit draft PR
+
+    # [] 1D Place Field Widget
+        # [X] Incorporate nelpy package into widget
+        # [] Add foreign group and sort controller to pick unit groups and ranges?
+        # [] Normalized firing rate figure?
+        # [] Add collapsed unit vizualization?
+        # [] Scale bar?
+        # [] Sort place cell tuning curves by peak firing rate position?
+        # [] Color palette control?
+
+# [] Dropdown that controls which unit
+
+# [x] Work in buttons / dropdowns / sliders to modify following parameters in place field calculation:
+# [] Different epochs
+# [x] Gaussian SD
+# [x] Speed threshold
+# [] Minimum firing rate
+# [] Place field thresh (% of local max)
+
+# Put widget rendering here
+def find_nearest(arr, tt):
+    """Used for picking out elements of a TimeSeries based on spike times
+
+    Parameters
+    ----------
+    arr
+    tt
+
+    Returns
+    -------
+
+    """
+    arr = arr[arr > tt[0]]
+    arr = arr[arr < tt[-1]]
+    return np.searchsorted(tt, arr)
+
+
+def smooth(y, box_pts):
+    """Moving average
+
+    Parameters
+    ----------
+    y
+    box_pts
+
+    Returns
+    -------
+
+    """
+    box = np.ones(box_pts) / box_pts
+    return np.convolve(y, box, mode='same')
+
+
+def compute_speed(pos, pos_tt, smooth_param=40):
+    """Compute boolean of whether the speed of the animal was above a threshold
+    for each time point
+
+    Parameters
+    ----------
+    pos: np.ndarray(dtype=float)
+        in meters
+    pos_tt: np.ndarray(dtype=float)
+        in seconds
+    smooth_param: float, optional
+
+    Returns
+    -------
+    running: np.ndarray(dtype=bool)
+
+    """
+    speed = np.hstack((0, np.sqrt(np.sum(np.diff(pos.T) ** 2, axis=0)) / np.diff(pos_tt)))
+    return smooth(speed, smooth_param)
```

Comment:
These types of functions should go in https://github.com/NeurodataWithoutBorders/nwb-jupyter-widgets/tree/master/nwbwidgets/analysis